pkg/agent/multicast/mcast_controller.go

```diff
@@ -0,0 +1,235 @@
+// Copyright 2021 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package multicast
+
+import (
+	"context"
+	"net"
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+
+	"antrea.io/antrea/pkg/agent/config"
+	"antrea.io/antrea/pkg/agent/interfacestore"
+	"antrea.io/antrea/pkg/agent/openflow"
+)
+
+type eventType uint8
+
+const (
+	groupJoin eventType = iota
+	groupLeave
+
+	// Use IGMPv3 for IGMP query message because it is the main version for IGMP.
+	queryVersion = 3
+
+	workerCount = 2
+)
+
+type mcastGroupEvent struct {
+	group net.IP
+	eType eventType
+	time  time.Time
+	iface *interfacestore.InterfaceConfig
+}
+
+type GroupMemberStatus struct {
+	group        net.IP
+	localMembers map[string]bool
+	lastProbe    time.Time
+	mutex        sync.RWMutex
+}
+
+type Controller struct {
+	ofClient      openflow.Client
+	nodeConfig    *config.NodeConfig
+	groupDetector *GroupDetector
+	groupEventCh  chan mcastGroupEvent
+	groupCache    cache.Indexer
+	mutex         sync.RWMutex
+	ctx           context.Context
+	cancelFunc    context.CancelFunc
+}
+
+func (m *Controller) Run(stopCh <-chan struct{}) {
+	// Install the PacketIn relevant flows to packetIn the IGMP messages and the generic Multicast forwarding flows to
+	// enable local Pods to work as the Multicast client.
+	m.ofClient.InstallMulticastInitialFlows(uint8(openflow.PacketInReasonMC))
+	// Periodically query Multicast Groups on OVS.
+	go wait.NonSlidingUntil(func() {
+		m.groupDetector.queryIGMP(net.IPv4zero, queryVersion)
+	}, time.Second*queryIntervalSeconds, stopCh)
+
+	go wait.NonSlidingUntil(m.clearStaleGroups, time.Second*queryIntervalSeconds, stopCh)
```

Comment:
I could move function `clearStaleGroups` into worker, but I feel it is a bit difficult to maintain a per worker cache for the groupCache if we want to support featuers like querying local Multicast groups and the Pods have join in each group. 

If we want to maintain a per worker cache, we should enure the `mcastGroupEvent` target to one Multicast group is processed by the same worker first. Then it will introduce extra work to query the Multicast groups if we want to support in antctl in future, we need to loop all wokers' groupCache to get a complete set for local groups.

Reply:
Even if I move function `clearStaleGroups` into worker, another goroutine except for the worker is possible to access the groupCache by calling function `checkLastMember`. That goroutine is to check if other members still exist in a group after receiving the IGMP leave message (IGMP v2 messages). 
https://github.com/antrea-io/antrea/pull/2652/files#diff-8e302e747c71db4b020ddf40f1debcb088af5aaea45266888f5f153c1be1b67bR164

Reply:
Got it.